### PR TITLE
UM02 - uToken Exchange Rate Drop

### PR DESCRIPTION
**Pull request by 0xtaf**
# uToken Exchange Rate Drop

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

### CODE REVIEW

**Review by Hyodar — COMMENTED** [review]
I made some comments, let me know what you think about them.

**Comment by 0xtaf** [comment]
> I made some comments, let me know what you think about them.

using bignumber.js made dealing with decimals easier. I updated all the parts you mentioned above 👍

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/fetcher.spec.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    return mockProvider.addCallTo(uTokenAddress, blockNumber, UTOKENS_IFACE, "UNDERLYING_ASSET_ADDRESS", {
+      inputs: [],
+      outputs: [underlyingAsset],
+    });
+  }
+
+  function createMockPriceCall(assetAddress: string, blockNumber: number, price: BigNumber) {
+    return mockProvider.addCallTo(TEST_CONFIG.umeeOracle, blockNumber, UMEE_ORACLE_PRICE_IFACE, "getAssetPrice", {
+      inputs: [assetAddress],
+      outputs: [price.toString()],
```
Comment (Hyodar) [inline Umee-Bots/uToken-Exchange-Rate-Drop/src/fetcher.spec.ts:47]:
```suggestion
      outputs: [price.toString(10)],
```

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+import { createFinding, calculateSeverity, AgentConfig } from "./utils";
+import BigNumber from "bignumber.js";
+
+const mockFetcher = {
+  getUnderlyingAssetAddress: jest.fn(),
+  getPrice: jest.fn(),
+  getReserveNormalizedIncome: jest.fn(),
+};
+
+const testUnderlyingAssetAddresses = [createAddress("0xua1"), createAddress("0xua2"), createAddress("0xua3")];
```
Comment (Hyodar) [inline Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.spec.ts:14]:
```suggestion
const testUnderlyingAssetAddresses = [createAddress("0x0a1"), createAddress("0x0a2"), createAddress("0x0a3")];
```
Minor, but better to keep addresses as hex strings.

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      if (difference.gte(pair.threshold)) {
+        const severity = calculateSeverity(difference, pair.threshold, pair.difference);
+        findings.push(createFinding(`${pair.uToken1}/${pair.uToken2}`, previousRatio, currentRatio, severity));
+      }
+    });
+
+    return findings;
+  };
+
+const getUTokenPrices = async (fetcher: Fetcher, underlyingAssets: any, block: string | number) => {
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.ts:54]: Maybe this function could be in `utils.ts`?

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+export function createFinding(
+  pair: string,
+  previousRatio: BigNumber,
+  currentRatio: BigNumber,
+  severity: FindingSeverity
+): Finding {
+  return Finding.from({
+    alertId: "UMEE-2",
+    name: "uToken exchange ratio drop",
+    description: `uToken pair of ${pair} has dropped.`,
```
Comment (Hyodar) [inline Umee-Bots/uToken-Exchange-Rate-Drop/src/utils.ts:31]:
```suggestion
    description: `${pair} rate had a significant drop from the previous block`,
```
Wdyt about something like this? Feel free to edit.

**Review by Hyodar — APPROVED** [review]
Great work! Approving because my comments are just about some details, so just please take a look at them when possible.

File: Umee-Bots/uToken-Exchange-Rate-Drop/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  - Fired when a uToken/uToken pair ratio is dropped largely
+  - Severity can be "info", "low", "medium", "high", "critical" depending on size of the drop
+  - Type is always set to "info"
+  - Metadata:
+    - `pair`: pair of uTokens
+    - `previousRatio`: previous ratio of the uToken pair before drop
+    - `currentRatio`: current ratio of the uToken pair after drop
+
+## Test Data
+
```
[the remark this answers is not in the document]
Comment (0xtaf) [reply Umee-Bots/uToken-Exchange-Rate-Drop/README.md:38]: makes sense. I'll do it

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/fetcher.spec.ts (hunk 1 above)
Comment (0xtaf) [reply Umee-Bots/uToken-Exchange-Rate-Drop/src/fetcher.spec.ts:47]: this is the ethers.BigNumber in order to simulate the return type so I'll keep this one

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/utils.ts (hunk 4 above)
Comment (0xtaf) [reply Umee-Bots/uToken-Exchange-Rate-Drop/src/utils.ts:31]: alright. more descriptive 👍

File: Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xtaf) [reply Umee-Bots/uToken-Exchange-Rate-Drop/src/agent.ts]: makes sense. done.

**Review by yehia67 — APPROVED** [review]

**Pull request merged**